test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -1446,6 +1482,25 @@ private synchronized void startAndPublishNodesAndClients(List<NodeAndClient> nod
         }
     }
 
+    private void rebuildUnicastHostFiles(Collection<NodeAndClient> newNodes) {
+        try {
+            List<String> discoveryFileContents = Stream.concat(nodes.values().stream(), newNodes.stream())
+                .map(nac -> nac.node.injector().getInstance(TransportService.class)).filter(Objects::nonNull)
+                .map(TransportService::getLocalNode).filter(Objects::nonNull).filter(DiscoveryNode::isMasterNode)
+                .map(n -> n.getAddress().toString())
+                .distinct().collect(Collectors.toList());
+            Set<Path> configPaths = Stream.concat(nodes.values().stream(), newNodes.stream())
+                .map(nac -> nac.node.getEnvironment().configFile()).collect(Collectors.toSet());
+            logger.info("configuring discovery with {} at {}", discoveryFileContents, configPaths);
+            for (final Path configPath : configPaths) {
+                Files.createDirectories(configPath);
+                Files.write(configPath.resolve(UNICAST_HOSTS_FILE), discoveryFileContents); // TODO do we need to do this atomically?
```

Comment:
NB open question: do we need to do this atomically?

Reply:
don't think so? We never call this concurrently for the same destination file?